panoply/datasource.py

```diff
@@ -18,13 +24,58 @@ def log(self, *msgs):
         if "logger" in self.options:
             self.options["logger"](msgs)
         else:
-            print msgs
+            print(msgs)
 
-    def progress(self, loaded, total, msg):
+    def progress(self, loaded, total, msg=''):
         self.fire("progress", {
             "loaded": loaded,
             "total": total,
             "msg": msg
         })
 
+# a decorator used to invalidate the access_token for oauth based data sources
+# this should be used on every method in the data source that fetches data
+# from the server (controlled by this oauth), and that needs to be invalidated
+def invalidate_token(refresh_url, callback=None,
+                     keys=['access_token', 'refresh_token']):
+    def _invalidate_token(f):
+        def wrapper(*args):
+            self = args[0]
+            try:
+                return f(*args)
+            except:
```

Comment:
You can have a generic function that will take care for refreshing the token, and let the user decide when to call it.(from the data source implementation itself)
It means that this function can't be a decorator, but I am not sure that such function should be a decorator.